pyat/at/physics/nonlinear.py

```diff
@@ -193,12 +235,165 @@ def gen_detuning_elem(ring: Lattice, orbit: Optional[Orbit] = None) -> Element:
         orbit, _ = find_orbit(ring)
     lindata0, _, _ = ring.linopt6(get_chrom=False, orbit=orbit)
     xsi = get_chrom(ring.radiation_off(copy=True))
-    r0, r1, x, q_dx, y, q_dy = detuning(ring.radiation_off(copy=True),
-                                        xm=1.0e-4, ym=1.0e-4, npoints=3)
-    nonlin_elem = Element('NonLinear', PassMethod='DeltaQPass',
-                          Betax=lindata0.beta[0], Betay=lindata0.beta[1],
-                          Alphax=lindata0.alpha[0], Alphay=lindata0.alpha[1],
-                          Qpx=xsi[0], Qpy=xsi[1],
-                          A1=r1[0][0], A2=r1[0][1], A3=r1[1][1],
-                          T1=-orbit, T2=orbit)
+    r0, r1, x, q_dx, y, q_dy = detuning(
+        ring.radiation_off(copy=True), xm=1.0e-4, ym=1.0e-4, npoints=3
+    )
+    nonlin_elem = Element(
+        "NonLinear",
+        PassMethod="DeltaQPass",
+        Betax=lindata0.beta[0],
+        Betay=lindata0.beta[1],
+        Alphax=lindata0.alpha[0],
+        Alphay=lindata0.alpha[1],
+        Qpx=xsi[0],
+        Qpy=xsi[1],
+        A1=r1[0][0],
+        A2=r1[0][1],
+        A3=r1[1][1],
+        T1=-orbit,
+        T2=orbit,
+    )
     return nonlin_elem
+
+
+def feeddown_polynomba(
+    polb: numpy.ndarray or tuple = (),
+    pola: numpy.ndarray or tuple = (),
+    xoffset: float = 0,
+    yoffset: float = 0,
+    verbose: bool = False,
+    debug: bool = False,
+) -> dict[str, numpy.ndarray]:
+    """
+    Return the feeddown due to a transverse offset.
+
+    Parameters:
+        polb: numpy array. PolynomB.
+        pola: numpy array. PolynomA.
+        xoffset: float. Default zero. Horizontal offset in meters.
+        yoffset: float. Default zero. Vertical offset in meters.
+        verbose: prints additional info.
+        debug: prints information about the feeddown polynom construction.
+
+    Returns:
+        Dictionary with PolynomB and PolynomA feeddown components.
+
+    Raises:
+        ValueError: if none of the two polynoms is passed.
+
+    Note:
+        If only one polynom is passed, it is assumed to be PolynomB.
+    """
+    # check debug and verbose flags only once
+    debugprint = print if debug else lambda *a, **k: None
+    verboseprint = print if verbose else lambda *a, **k: None
+
+    # verify polynoms length
+    maxorda = len(pola)
+    maxordb = len(polb)
+    if maxorda == 0 and maxordb == 0:
+        raise ValueError("At least one polynom is needed")
+    maxord = max(maxorda, maxordb)
+    debugprint(f"maxord={maxord},maxorda={maxorda},maxordb={maxordb}")
+
+    debugprint("Padding polynoms")
+    polbpad = numpy.pad(polb, (0, maxord - maxordb), "constant", constant_values=(0, 0))
+    polapad = numpy.pad(pola, (0, maxord - maxorda), "constant", constant_values=(0, 0))
```

Comment:
It seems that even if the input polynom is empty you are padding it with zeros and going through the whole calculation.
Maybe this could be avoided?

In this case you could even avoid to raise exception: an empty input polynom returns an empty output polynom

Up to you...

Reply:
Code reordered to avoid padding zero-length polynoms